src/LinearSolverTypes.f90

```diff
@@ -1615,228 +1662,216 @@ SUBROUTINE solveBiCGSTAB(solver)
+!> @param thisPC The preconditioner object to use on the system
+!>
+!> This subroutine solves the Iterative Linear System using the restarted GMRES method
+!>
+SUBROUTINE solveGMRES(thisLS,thisPC)
+  CLASS(LinearSolverType_Iterative),INTENT(INOUT) :: thisLS
+  CLASS(PreconditionerType),INTENT(INOUT),OPTIONAL :: thisPC
+  INTEGER(SIK) :: nIters,outerIt
+  REAL(SRK) :: tol,norm_b
+  CLASS(VectorType),ALLOCATABLE :: u
+
+  IF (thisLS%nRestart > thisLS%A%n) thisLS%nRestart = thisLS%A%n
+  thisLS%iters = 0
+  norm_b=BLAS_nrm2(thisLS%b)
+  IF (norm_b .APPROXEQ. 0.0_SRK) THEN
+    CALL thisLS%X%set(0.0_SRK)
+    thisLS%residual = 0.0_SRK
+    thisLS%iters = 1
+    RETURN
+  ENDIF
 
-  n=0
-  !Set parameter list for vector
-  CALL pList%add('VectorType -> n',solver%A%n)
-  n=solver%A%n
-  m=MIN(solver%nRestart,n)
-  CALL u%init(pList)
-  CALL pList%clear()
-  CALL solver%getResidual(u)
-  CALL LNorm(u%b,2,beta)
-  solver%iters=0
+  CALL VectorResembleAlloc(u,thisLS%x)
+  CALL thisLS%getResidual(residVec=u)
+  thisLS%residual=BLAS_nrm2(u)
+  tol=MAX(thisLS%absConvtol,thisLS%relConvTol*thisLS%residual)
+  tol=tol*norm_b
+  IF (thisLS%residual .APPROXLE. tol) THEN
+    thisLS%iters = 1
+    RETURN
+  ENDIF
 
-  IF(beta > EPSILON(0.0_SRK)) THEN
-    ALLOCATE(v(n,m+1))
-    ALLOCATE(R(m+1,m+1))
-    ALLOCATE(w(n))
-    ALLOCATE(c(m+1))
-    ALLOCATE(s(m+1))
-    ALLOCATE(g(m+1))
-    ALLOCATE(y(m+1))
-    v(:,:)=0._SRK
-    R(:,:)=0._SRK
-    w(:)=0._SRK
-    c(:)=0._SRK
-    s(:)=0._SRK
-    g(:)=0._SRK
-    y(:)=0._SRK
-    tol=solver%absConvTol*beta
-
-    v(:,1)=-u%b/beta
-    h=beta
-    phibar=beta
-#ifdef FUTILITY_DEBUG_MSG
-      WRITE(668,*) '         GMRES-NOPC',0,ABS(phibar)
-#endif
-    !Iterate on solution
-    DO it=1,m
-      CALL BLAS_matvec(THISMATRIX=solver%A,X=v(:,it),BETA=0.0_SRK,Y=w)
-      h=BLAS_dot(n,w,1,v(:,1),1)
-      w=w-h*v(:,1)
-      t=h
-      DO k=2,it
-        h=BLAS_dot(n,w,1,v(:,k),1)
-        w=w-h*v(:,k)
-        R(k-1,it)=c(k-1)*t+s(k-1)*h
-        t=c(k-1)*h-s(k-1)*t
-      ENDDO
-      CALL LNorm(w,2,h)
-      !WRITE(*,*) "h = ", h
-      IF(h > 0.0_SRK) THEN
-        v(:,it+1)=w/h
+  SELECT TYPE(u); CLASS IS(NativeVectorType)
+    DO outerIt=1,CEILING(thisLS%maxIters*1.0/thisLS%nRestart)
+      IF(PRESENT(thisPC)) THEN
+        CALL solveGMRES_partial(thisLS,u,tol,nIters,thisPC)
       ELSE
-        v(:,it+1)=0.0_SRK*w
+        CALL solveGMRES_partial(thisLS,u,tol,nIters)
       ENDIF
-      !Set up next Given's rotation
-      IF(t >= 0.0_SRK) THEN
-        temp=SQRT(t*t+h*h)
-      ELSE
-        temp=-SQRT(t*t+h*h)
+      thisLS%iters=thisLS%iters+nIters
+      IF (thisLS%residual .APPROXLE. tol) THEN
+        EXIT
       ENDIF
-      c(it)=t/temp
-      s(it)=h/temp
-      R(it,it)=temp
-      g(it)=c(it)*phibar
-      phibar=-s(it)*phibar
-#ifdef FUTILITY_DEBUG_MSG
-      WRITE(668,*) '         GMRES-NOPC',it,ABS(phibar)
-#endif
-      IF(ABS(phibar) <= tol) EXIT
     ENDDO
```

Comment:
The last version had a few lines right inside this ENDDO:
```
-      CALL thisLS%getResidual(residVec=u)
-      IF(PRESENT(thisPC)) CALL thisPC%apply(u)
-      thisLS%residual = BLAS_nrm2(u)
```
These deletions were intentional?  Why aren't they needed?

Reply:
These were moved to the end of the solveGMRES_partial routine. Ultimately it doesn't really matter so much whether they're put here or there, since either way they get called at the end of every restart.

Reply:
Ok, works for me.  I missed it in the other location.